service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/VmExtensionHelper.java

```diff
@@ -0,0 +1,190 @@
+package bio.terra.workspace.service.resource.controlled.cloud.azure.vm;
+
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.workspace.common.exception.AzureManagementException;
+import bio.terra.workspace.common.exception.AzureManagementExceptionUtils;
+import bio.terra.workspace.common.utils.AzureUtils;
+import bio.terra.workspace.generated.model.ApiAzureVmCreationParameters;
+import bio.terra.workspace.generated.model.ApiAzureVmCustomScriptExtension;
+import com.azure.core.management.exception.ManagementException;
+import com.azure.resourcemanager.compute.ComputeManager;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Knows how to install and manage a custom script extension on a VM */
+public class VmExtensionHelper {
+  private static final Logger logger = LoggerFactory.getLogger(VmExtensionHelper.class);
+
+  /**
+   * Installs a custom script extension on a VM if the extension is configured in the supplied
+   * creation parameters
+   *
+   * @param creationParameters
+   * @param vmId
+   * @param computeManager
+   * @return
+   */
+  public StepResult maybeInstallExtension(
+      ApiAzureVmCreationParameters creationParameters, String vmId, ComputeManager computeManager) {
+    if (creationParameters == null || creationParameters.getCustomScriptExtension() == null) {
+      logger.info("No custom script extension to install on VM {}, skipping", vmId);
+      return StepResult.getStepResultSuccess();
+    }
+
+    logger.info("Checking if custom script extension is already installed on VM {}", vmId);
+    var extensionStatus =
+        this.checkExtensionInstalled(
```

Comment:
It's good that there are specific API in place for this and we don't need to manage flow/idempotency with try/catch.